src/models/logits.cpp

```diff
@@ -9,33 +9,40 @@ Logits::Logits(const Model& model, State& state)
       state_{state},
       shape_{static_cast<int64_t>(state_.params_->batch_size) * state_.params_->search.num_beams, state_.params_->sequence_length, state_.params_->vocab_size},
       type_{model_.session_info_->GetOutputDataType(model_.config_->model.decoder.outputs.logits)} {
-  // DML doesn't support on-device scoring yet, so fall back to the CPU
-  auto& allocator = model_.device_type_ == DeviceType::DML ? model_.allocator_cpu_ : *model_.allocator_device_;
-  auto logits_tensor = OrtValue::CreateTensor(allocator, shape_, type_);
+  auto logits_tensor = OrtValue::CreateTensor(*model_.allocator_device_, shape_, type_);
   if (type_ == Ort::TypeToTensorType<float>::type)
     value32_ = std::move(logits_tensor);
   else
     value16_ = std::move(logits_tensor);
 
-  if (model_.device_type_ == DeviceType::CUDA && model_.use_cuda_graph_) {
-    size_t max_beam_batch_size = static_cast<size_t>(model_.config_->search.num_beams) * model_.max_batch_size_;
+  if (state_.GetCapturedGraphInfo()) {
     if (type_ == Ort::TypeToTensorType<float>::type) {
-      sb_logits32_ = std::make_unique<StaticBuffer>(model_.allocator_device_, max_beam_batch_size);
+      sb_logits32_ = state_.GetCapturedGraphInfo()->sb_logits32_.get();
     }
     if (type_ == Ort::TypeToTensorType<Ort::Float16_t>::type) {
-      sb_logits16_ = std::make_unique<StaticBuffer>(model_.allocator_device_, max_beam_batch_size);
+      sb_logits16_ = state_.GetCapturedGraphInfo()->sb_logits16_.get();
     }
   }
 }
 
 RoamingArray<float> Logits::Get() {
   size_t element_count = shape_[0] * shape_[1] * shape_[2];
-  // DML doesn't support on-device scoring yet, so fall back to the CPU
-  auto& allocator = model_.device_type_ == DeviceType::DML ? model_.allocator_cpu_ : *model_.allocator_device_;
 
   // Convert from float16 to float32 if necessary
-  if (type_ == Ort::TypeToTensorType<Ort::Float16_t>::type)
-    ConvertFp16ToFp32(allocator, *value16_, value32_, model_.device_type_, model_.cuda_stream_);
+  if (type_ == Ort::TypeToTensorType<Ort::Float16_t>::type) {
+#if USE_DML
+    DmlHelpers::DmlCastInputToOutput(
```

Comment:
Could this be added to ConvertFp16ToFp32, it already works for CPU & CUDA, DML shouldn't be a special case here.